RFC: Allow (outer) cfg-attributes on elements of tuple types and tuple (struct) patterns

recatek opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3532)*

Let's make it more elegant to conditionally compile tuple type declarations and pattern matches by allowing (outer) cfg-attributes directly on their elements.

[Rendered](https://github.com/recatek/rfcs/blob/cfg-attribute-in-tuple-type/text/3532-cfg-attr-in-tuple-tys-and-pats.md)

recatek committed:
Create 0000-cfg-attribute-in-tuple-type.md

recatek committed:
Update 0000-cfg-attribute-in-tuple-type.md

recatek committed:
Adding links to RFC #3399

recatek committed:
Update 0000-cfg-attribute-in-tuple-type.md

recatek committed:
Assigning RFC number

ahicks92 commented:
I agree that this makes the language consistent and so it's probably worth doing, however the idea of conditionally including or excluding elements from tuples or tuple structs fills me with dread because of all the cases where maybe you get two different 3-tuples and the types happen to line up or something.  Maybe it's worth finding a use case for the motivation that's not a generic "maybe this shouldn't be exceptional".

recatek commented:
My particular use case is for working on [gecs](https://docs.rs/gecs/latest/gecs/), a compile-time entity component system that relies on tuples to work properly. There are a number of features I can't easily add to this system while also supporting conditional compilation. Conditional compilation in this case is important because you may have components that only exist on the client, or on the server, or only exist in debug builds, and so on. While I agree you could accidentally misalign types with this method, you already could do that with the combinatorial alternative of redefining each tuple type -- something that is likely to happen in a macro due to the amount of boilerplate involved.

clarfonthey commented:
I like this change, although my biggest concern is that because tuples are indexed by numbers, this change technically makes the exact indices of fields… weird.

One thing that might be a bit less disruptive is extending the tuple struct expression syntax to work with defining structs. Specifically, this syntax:

```rust
struct Tuple(&'static str, bool);

// you can just do this
const TUP: Tuple = Tuple { 0: "hello", 1: true };
```

So, the proposal would be that you could define structs like this:

```rust
struct Tuple {
    0: &'static str,
    1: bool,
}
```

And then, you could do weird things like:

```rust
struct Tuple {
    0: &'static str,
    #[cfg(one_thing)]
    1: bool,
    #[cfg(another_thing)]
    1: u8,
    #[cfg(third_thing)]
    2: i32,
}
```

And this way, you know what the indices of the fields are, and you also have the guarantee where, if this causes a gap in the indices (for example, if `one_thing` and `another_thing` don't apply when `third_thing` does, removing field 1), instead of just shifting the elements along, it just throws an error.

We could also add an auto-applicable suggestion to convert from the proposed syntax, where `#[cfg(thing)]` attributes are added to individual fields, to this syntax.

recatek commented:
I'm not sure that's all that functionally different from defining structs as you can now like so:

```rust
struct Something {
    a: SomeTypeA,
    b: SomeTypeB,
    #[cfg(feature = "foo")]
    c: SomeTypeC,
}
```

Just using `a`, `b`, and `c` instead of `0`, `1`, and `2` for accessors. Unfortunately that doesn't cover the use case in question here. The reason I ran into this issue is that I need to use tuples specifically to construct archetype storage, in a way similar to how bevy constructs its queries (see [here](https://promethia-27.github.io/dependency_injection_like_bevy_from_scratch/) and [here](https://github.com/bevyengine/bevy/discussions/9816)). Queries use tuples here to get variadic-like behavior, and doing so with tuples would allow me to construct ECS archetypes like so:

```rust
type SomeArchetype = EcsArchetype<(
    ComponentA,
    ComponentB,
    ComponentC,
    #[cfg(feature = "server")]
    ServerOnlyComponent,
)>;
```

Except `#[cfg]` is not supported in that position. What I *could* do legally now is

```rust
#[cfg(not(feature = "server"))]
type SomeArchetype = EcsArchetype<(
    ComponentA,
    ComponentB,
    ComponentC,
)>;

#[cfg(feature = "server")]
type SomeArchetype = EcsArchetype<(
    ComponentA,
    ComponentB,
    ComponentC,
    ServerOnlyComponent,
)>;
```

But that doesn't scale well if you have multiple features at play. In the short term I will likely create a macro to do this, but it's far from ideal -- supporting `#[cfg]` in this position the way it is for tuple initialization would make much more sense than these workarounds.

recatek commented:
Also, as an additional consistency note, this is already supported in tuple-based structs. The following is valid code:

```rust
pub struct SomeStruct(i32, #[cfg(all())] u32);

fn main() {
    let thing = SomeStruct(0, 1);
    println!("{}", thing.1)
}
```

ahicks92 commented:
I think that the interesting use case is gecs.  I haven't dug into the code but "these are the special things about tuples" would go a long way here.  Again though, this is probably worth doing just for consistency.

I will have to look into gecs more for personal projects; if you have somehow solved being able to jump into Lua scripts in the middle of systems that may solve many problems. But that's way off topic. Still, neat idea.

recatek commented:
Fair warning, gecs does some truly dreadful things (which I'm trying to work out better ways to do as Rust improves its metaprogramming stories over time). For tuples specifically, this kind of soft variadism, if that's a word, is pretty common in ECS libraries. I mentioned bevy's queries above, but other libraries like [hecs](https://docs.rs/hecs/latest/hecs/) and [legion](https://docs.rs/legion/latest/legion/) do similar things. You can't really do this kind of logic with a struct because it relies on the type being easy to anonymously construct on the fly.

The core conceit is that you want to be able to define structs/traits in the form of `Something<T>`, where `T` is in fact a collection of types described by a tuple, and all of the methods and dependent types are pre-defined for all tuples of arity 1 through, say, 16. You can't do this pre-definition if you require the end user to define an explicitly named struct, so you *must* use tuples here.

ahicks92 commented:
Yes, I know and have used the technique.  Warp takes it even further: if you define the right traits you can concatenate and flatten tuples as well.

I'm just saying that advocating for this, writing down such use cases would be a good idea.  I don't like it, but I agree it should be in the language.  I guess that's a weird stance.  But my feedback is primarily about readability.  I don't think most people consuming this are going to see us vaguely gesture at ECS and warp and go "yeah right type metaprogramming" and I know about such things and still required that you do so to remember they exist.  The real problem of course is that Rust doesn't have a variadics story but that's neither here nor there.

recatek committed:
Adding more information about use cases in the motivation section

recatek commented:
I've added a motivation section walking through the interaction between cfg-attributes and walking through the construction of a hypothetical entity archetype for multiplayer asteroids. Does that illustrate the need well, do you think?

recatek committed:
Clarifying interaction with macros and combinatorial expansion

recatek committed:
Fixing up example cfg-attributes

ahicks92 commented:
Yeah, that's pretty good I think.

recatek committed:
Minor fixes for code formatting

recatek commented:
For added context I started a discussion about the broader question of cfg-attribute rule consistency on IRLO [here](https://internals.rust-lang.org/t/a-discussion-about-addressing-inconsistencies-with-cfg-attributes/19928). I also put together a [HackMD page](https://hackmd.io/@recatek/S1NO5ZXHT) with all of the comma-terminated fragments in Rust that I'm aware of (suggested additions welcome!) and whether you can or can't use a cfg-attribute on them. There's a lot of inconsistency that I'd like to try to resolve, either case-by-case or just all up if possible.

kwfn commented on text/3532-cfg-attribute-in-tuple-type.md:
````diff
@@ -0,0 +1,323 @@
+)>;
+```
+
+Additionally, we need some components to handle serializing the network state, performing dead reckoning, and sending that information to the client from the server. So we will add a `StateStorageComponent` and a `DeltaCompressionComponent`, and restrict those to the server, since the client does not perform these calculations and we want to avoid giving clients this information to help defeat, o ...
````
```suggestion
Additionally, we need some components to handle serializing the network state, performing dead reckoning, and sending that information to the client from the server. So we will add a `StateStorageComponent` and a `DeltaCompressionComponent`, and restrict those to the server, since the client does not perform these calculations and we want to avoid giving clients this information to help confound, or at least delay, cheaters.
```

recatek committed:
Update text/3532-cfg-attribute-in-tuple-type.md

Co-authored-by: oliver <151407407+kwfn@users.noreply.github.com>

recatek committed:
Update 3532-cfg-attribute-in-tuple-type.md

recatek commented:
After playing around with this a bit, it occurs to me that this should be bundled with allowing cfg-attributes in pattern matching on said tuples as well, to make it easier to break down these conditionally-described tuples. In the `hecs` or `bevy`-like ECS example, this would look like:

```rust
type MovementQuery = (Position, Rotation, #[cfg(server)] Authority);
for (position, rotation, #[cfg(server)] authority) in ecs.query::<MovementQuery>().iter_mut() {
    let _output = update_movement(position, rotation);
    
    #[cfg(server)]
    update_authority(position, rotation, _output);
}
```

I think I'll amend the RFC to include these.

recatek committed:
Adding pattern matching

weiznich commented:
@traviscross What would be needed to move this forward? This would be quite nice to have for diesel as well.

steffahn commented:
I’m comparing with other cases of “positional arguments” in type contexts, and see they seem disallowed, too. In particular:


```rs
fn foo(x: impl FnOnce(i32, #[cfg(all())] i32)) {}
```
```
   Compiling playground v0.0.1 (/playground)
error: `Trait(...)` syntax does not support attributes in parameters
 --> src/lib.rs:2:28
  |
2 | fn foo(x: impl FnOnce(i32, #[cfg(all())] i32)) {}
  |                            ^^^^^^^^^^^^^ help: remove the attributes
```

```rs
struct S<T, U = T> (T, U);

fn foo(x: S<(), #[cfg(any())] i32>) {}
```
```
error: attributes cannot be applied to generic arguments
 --> src/lib.rs:4:17
  |
4 | fn foo(x: S<(), #[cfg(any())] i32>) {}
  |                 ^^^^^^^^^^^^^ attributes are not allowed here
```

and I don’t see this changing with this RFC. So I think it would *decrease* consistency, wouldn’t it?

recatek commented:
I maintain a [HackMD page](https://hackmd.io/@recatek/S1NO5ZXHT) listing all of the comma-terminated fragments in Rust and whether or not they can be decorated with cfg-attributes. The vast majority of them can be. The generic arguments you mention are one of the few outliers, and I intend to RFC support for those eventually as I believe all comma-terminated fragments should be supported. Currently I'm focusing on one category at a time, and was waiting on [RFC 3399](https://rust-lang.github.io/rfcs/3399-cfg-attribute-in-where.html) before doing another generics cfg-attribute RFC.

The Trait(...) syntax for `impl FnOnce(i32, #[cfg(false)] u32)` one is new to me, and I'll add it to that list. I believe it should also be supported given that `fn foo(f: fn(i32, #[cfg(false)] u32)) {}` _is_ supported.

Generally, I don't expect Rust to remove support for existing cfg-attribute locations, so the direction to move for greater consistency is adding them. Comma-terminated fragments are self-contained things that I believe are fairly straightforward to functionally omit in any place where they appear, and so being able to control that with conditional compilation makes the most sense to me.

steffahn commented:
> I believe it should also be supported given that fn foo(f: fn(i32, #[cfg(false)] u32)) {} is supported.

Oh that *is* an interesting datapoint, thanks for pointing this one out. Arguably, the arguments of `fn(…) -> T` pointers is actually conceptually somewhat *closer* to the arguments of tuple types than it is to the two example I brought up. In any case, I believe it’s probably worth mentioning some (all?) of these cases for “positional arguments in type expressions” in the RFC text too.

---

Looking at the future-possibilities section, I’m not sure what’s the precedent of linking to a hackmd document; this could have downsides of making the RFC harder to comprehend for future readers, in case the HackMD sees further updates in the mean-time. Like a sentence like 
> “If this RFC is accepted, only two reasonable use cases for cfg-attributes on comma-terminated fragments would remain uncovered.”

is hard to make precise sense of if that additional context could get altered (or even disappear) in the future.

steffahn commented:
@recatek in your HackMD, you probably want the `fn foo(f: fn(i32, #[cfg(false)] i32)) {} // OK` example – i.e. the case of *items in arguments list of **function pointer types*** [-> [MaybeNamedParam](https://doc.rust-lang.org/1.94/reference/types/function-pointer.html#railroad-MaybeNamedParam)] – under “Can be decorated”, right?

---

Actually… another interesting detail might be that the `fn(…) -> T` syntax is also special in terms of accepting optionally *named* arguments. With names, it’s looking like:
```rs
fn foo(f: fn(x: i32, #[cfg(false)] y: i32)) {} // OK
```

In principle, this could of course could possibly be a counter-argument again, i.e. “`fn` pointers only allow this because their syntax is *also* special in this other way”.

programmerjake commented:
> Actually… another interesting detail might be that the `fn(…) -> T` syntax is also special in terms of accepting optionally _named_ arguments.

I think we should allow named arguments in `Fn(...)` syntax too, it makes much better documentation on what the arguments for callbacks mean. e.g.:
```rust
fn get_movie_and_walk_data_when_done(callback: impl FnOnce(steps: u32, with_popcorn: bool, movie_length: Duration, movie_name: &str)) { ... }
```

recatek commented:
I'd be happy to migrate the HackMD to a GitHub repo if we want something more durable and versioned contemporaneously with the RFC itself. Either that, or I could embed it as an appendix -- can an RFC have an appendix? Also, does editing the RFC text at this point affect the FCP?

recatek committed:
Adding appendix of cfg-attribute usage on comma-terminated fragments

Adding an appendix enumerating the current state of (from https://hackmd.io/@recatek/S1NO5ZXHT) of cfg-attribute usage on all comma-terminated fragments I'm aware of

fmease commented on text/3532-cfg-attribute-in-tuple-type.md:
```diff
@@ -0,0 +1,564 @@
+- Feature Name: `cfg_attribute_in_tuple_type`
```
This RFC also proposes allowing attributes in tuple patterns but this feature name and the PR title don't reflect that fact.

Could you change the PR title to something like *RFC: Allow (outer) cfg-attributes on elements of tuple types and tuple (struct) patterns* and the feature name to something akin to `cfg_attr_in_tuple_tys_and_pats`?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3532/845d609e0025d07c8cb0ca017d942bdbe9a18484..afbcb07e789651fc3abed2955f6f0444758c3b7d)*

recatek replied:
Done and done!

recatek committed:
Renaming to cfg-attr-in-tuple-tys-and-pats

recatek committed:
... .md

ehuss commented on text/3532-cfg-attr-in-tuple-tys-and-pats.md:
```diff
@@ -0,0 +1,564 @@
+
+This RFC proposes changing the syntax of the `TupleType` (See [[type.tuple.syntax]](https://doc.rust-lang.org/stable/reference/types/tuple.html#r-type.tuple.syntax)) to include `OuterAttribute*` before each occurrence of `Type`. These attributes can decorate each individual type (up to the comma or closing paren). Similarly, for pattern matching, this RFC proposes adding `OuterAttribute*` before  ...
+
+In practice, at least within the scope of this RFC, only cfg-attributes need to be supported in these new `OuterAttribute`s.
```
This would be unusual to explicitly forbid the lint-level attributes. They are generally supported anywhere that an attribute is allowed. It probably won't be too common, but there could be situations like `#[allow(deprecated)]`.

Also of note:
- `diagnostic` attributes are allowed anywhere, so would be allowed here.
- Tool attributes are allowed anywhere, so would be allowed here.
- Doc attributes require special attention in the compiler to specify where they are allowed which is easy to miss, and possible for a naive user to just try adding (and then get confused when it does nothing). I'm not saying that needs to be supported, but more so that the implementation needs to be careful to explicitly check it which I think has a high risk for being forgotten (its exhaustive Target checking is currently disabled).
- Similarly, rustfmt (for things like `rustfmt::skip`) is often forgotten, and when it just doesn't work (with no warnings) causes users confusion. (This seems unlikely, but who knows.)


*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3532/845d609e0025d07c8cb0ca017d942bdbe9a18484..35e78095438efb78cae1440f6b2251e38562afdc)*

recatek replied:
Just to clarify, I'm not saying that we shouldn't allow these things (if anything, it would be great if we could!), just that allowing them isn't what this RFC is focused on. Whether or not they're supported could be an implementation detail or a different discussion -- this RFC is neutral to the inclusion/exclusion of anything in that position other than cfg-attributes.

ehuss replied:
I don't want to belabor the point too much since it is relatively minor. However, the reason for being precise when making a language change is specifically to avoid leaving questions until implementation. This helps avoid confusion, people having inconsistent expectations, unintentionally missing changes that should happen, needing last-minute changes, having broken or inconsistent behavior, and potentially uncovering aspects that the lang team needs to consider.

I realize the intent of this RFC is to focus on adding `cfg` support, but when making one change in the language, it often has ramifications in other parts of the language. I think it is important when making language changes to consider those ramifications.

So I would recommend changing this to say something like:

The following attributes will be allowed in these attribute positions:

- `cfg` and `cfg_attr`
- The lint-level attributes (because they are allowed in all attribute positions).
- `diagnostic` attributes (because they are allowed in all attribute positions, but generate a warning).
- Tool attributes (which are not restricted by rustc).

And in particular, the `doc` attribute and doc comments should *not* be allowed.

recatek replied:
Sure, I'll add a section about this for clarity where the RFC talks about spec changes, using your suggested allow/deny list. One point of clarification:

> And in particular, the `doc` attribute and doc comments should *not* be allowed.

That means this outcome:

```rust
pub struct Foo(
    /// A -- OK
    u32, 
    /// B -- OK
    u32,
);

pub struct Bar((
    /// A -- ERROR
    u32, 
    /// B -- ERROR
    u32,
));
```

I *think* the answer to this is that this is fine, but I just wanted to double check.

ehuss replied:
That seems like it should be fine to me. If for some reason we want to support it in the future, it can be added later.

recatek replied:
I've added this to the RFC body under a "What about other attributes?" subsection of the guide-level explanation.

recatek committed:
Adding a "What about other attributes?" section to the guide-level explanation
